starter/privileged_service_server.cc

```diff
@@ -0,0 +1,146 @@
+    size_t header_size = std::max(msg_len, sizeof(Response));
+    char *buf = msg.buf + header_size;
+    size_t buf_size = sizeof(msg) - header_size;
+    size_t value_len = 0; // set below to the actual length of the value to be returned
+    int rc = 0;
+    int fd_out = -1; // set below when 'rc' is a file descriptor
+
+    switch (msg.hdr.msg_type_) {
+    case TYPE_DUMP_REQUEST: // msg size == header size
+      value_len = dump_capabilities(CAP_EFFECTIVE, buf, buf_size);
+      break;
+    case TYPE_BPF_OPEN_REQUEST: // msg size == header size + path length
+      // zero terminate path name
+      msg.bpf_open_req.path_[msg_len - sizeof(msg.bpf_open_req)] = '\0';
+      {
+	union bpf_attr attr = {};
+	attr.pathname = uintptr_t(msg.bpf_open_req.path_);
+	fd_out = rc = ::syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+      }
+      break;
+    case TYPE_BPF_LOOKUP_REQUEST: // key_size = msg_len - sizeof msg.bpf_lookup_req
+      // require at least one byte key
+      if (msg_len < sizeof(msg.bpf_lookup_req) + 1) {
+	fprintf(stderr, "received truncated bpf lookup request (%zd bytes), skipping\n", msg_len);
+	rc = -1;
+	errno = EINVAL;
+	break;
+      }
+      value_len = msg.bpf_lookup_req.value_size_;
+      // Make sure the value fits into available space
+      if (buf_size < value_len) {
+        rc = -1;
+        errno = EINVAL;
+      } else {
+        union bpf_attr attr = {};
+        attr.map_fd = uint32_t(fd_in);
+        attr.key = uintptr_t(msg.bpf_lookup_req.key_);
+        attr.value = uintptr_t(buf);
+        rc = ::syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+      }
+      if (rc != 0) {
+        value_len = 0;
+      }
+      break;
+    case TYPE_SETSOCKOPT32_REQUEST: // msg_len == sizeof msg.setsockopt_req
+      if (msg_len < sizeof(msg.setsockopt_req)) {
+	fprintf(stderr, "received truncated setsockopt request (%zd bytes), skipping\n", msg_len);
+	rc = -1;
+	errno = EINVAL;
+	break;
+      }
+      rc = ::syscall(__NR_setsockopt, fd_in, msg.setsockopt_req.level_,
+		     msg.setsockopt_req.optname_, &msg.setsockopt_req.optval_,
+		     sizeof(msg.setsockopt_req.optval_));
+      break;
+    default:
+      fprintf(stderr, "Unexpected privileged call type: %d\n", msg.hdr.msg_type_);
+      rc = -1;
+      errno = EINVAL;
+    }
+
+    // Close the received file descriptor
+    if (fd_in != -1) {
+      ::close(fd_in);
+    }
+
+    // Form the response in place
+    msg.response.hdr_.msg_type_ = TYPE_RESPONSE;
+    if (fd_out != -1) {
+      // Pass a poitive but invalid fd in ret_, to be replaced with the passed
+      // fd by the receiver.
+      msg.response.ret_ = {INT_MAX, 0};
+    } else {
+      msg.response.ret_ = {rc, rc != -1 ? 0 : errno};
+    }
+    size = send_fd_msg(&msg, sizeof(msg.response), buf, value_len, fd_out);
+    if (size < ssize_t(sizeof(msg.response) + value_len)) {
+      perror("sendmsg");
+    }
+
+    // Close the sent file descriptor
```

Comment:
nit: `sending file descriptor`?

Reply:
Same here, when opening a bpf map, the opened file descriptor is sent in recvmsg control channel and we need to close the sent file descriptor, as otherwise the bpf map would stay open due to the server side keeping an open file descriptor to it.